src/shared/components/PillTag/PillTag.tsx

```diff
@@ -3,19 +3,67 @@ import _ from 'lodash';
 import styles from './styles.module.scss';
 import { If } from 'react-if';
 import contrast from 'contrast';
-import { computed, makeObservable } from 'mobx';
+import { computed, makeObservable, observable } from 'mobx';
+import { getBrowserWindow } from 'cbioportal-frontend-commons';
+import classnames from 'classnames';
+import { observer } from 'mobx-react';
+import { StudyViewPageStore } from 'pages/studyView/StudyViewPageStore';
 
 export interface IPillTagProps {
-    content: JSX.Element | string;
+    content: string | { uniqueChartKey: string; element: JSX.Element };
     backgroundColor: string;
     infoSection?: JSX.Element | null;
     onDelete?: () => void;
+    store: StudyViewPageStore;
 }
 
+export type PillStoreEntry = {
+    /**
+     * Unique filter key
+     */
+    key: string;
+
+    /**
+     * Called in hesitate mode, when filter updates are submitted
+     */
+    onDeleteCallback?: () => void;
+};
+
+export type PillStore = {
+    [content: string]: PillStoreEntry;
+};
+
+/**
+ * Keep track of submitted and 'hesitant' (i.e. queued or non-submitted)
+ * filters using a unique key. When a delete is queued, a hesitant store entry
+ * also contains an onDeleteCallback, which will be executed on submit.
+ *
+ * Filters can be of different shapes, and they are created
+ * in a number of different places, so we use these two global stores
+ * to keep track of hesitant (queued) filters and submitted filters.
+ *
+ * TODO: A better solution would be a uniform list of filters
+ *  but this would require some refactoring.
+ */
+getBrowserWindow().hesitantPillStore = {} as PillStore;
```

Comment:
why not put these in the StudyView context?

Reply:
also, is "hesitant" a typo?

Reply:
Fixed in [60e31e2](https://github.com/cBioPortal/cbioportal-frontend/pull/4502/commits/60e31e2d279ad9e2f6e6a0ff18bac2d4199c64f6) (pill stores are now in study view store, and study view store is passed to pills using the context)